### DISCUSSION TITLE
UUID matching

### DISCUSSION THREAD

**Discussion by leaanthony** [alecthomas/participle#168] [Q&A]
Hey Alec! Awesome project and easily the best parser library I've used. Just wandering the best way to parse a UUID? The default lexer is perfect for my use case, apart from this one thing. Is it possible to extend the default lexer or would I need to recreate the default lexer and add my specific rule? Cheers!
EDIT: I've tried consuming all tokens in a greedy fashion and this worked great until an 'e' was present, then I got a "exponent has no digits" error, which appears to be an error thrown by the Go scanner.

**Comment by alecthomas** [comment]
Glad you enjoy using it :)

There's no reasonable way to extend it, but you can construct one via an existing text/scanner. There may be a way to configure a text/scanner to do this? If not you'll need to roll your own, which is not too hard.

**Comment by alecthomas** [comment]
Check out the `_examples` folders for some custom lexers.

**Comment by leaanthony** [comment]
Thanks for the quick reply Alec. Is the default lexer available for extending? I'll be honest, I've no idea where to start to recreate it :-)

**Comment by alecthomas** [comment]
You want [this function](https://pkg.go.dev/github.com/alecthomas/participle/v2@v2.0.0-alpha6/lexer#LexWithScanner). It depends on whether the text/scanner can be modified to do what you want. It has some configurability.

**Comment by leaanthony** [comment]
Thanks. I'm happy to just greedily consume the uuid and tried using the `@~` syntax but I'm getting `unexpected token ~`. 
Thing to parse: `uuid(76eceacb-0462-4d94-ac83-423a6793775e)`
Current attempt: `'uuid' '(' @~')' ')'`

**Comment by alecthomas** [comment]
If `~` is a literal you will need to quote it the same way you've done with `uuid`, `(` and `)`, eg. `@'~'`

**Reply by leaanthony** [reply]
I was using the tilda as suggested in the README: 
![image](https://user-images.githubusercontent.com/1943904/127732246-9c1ca279-fbda-4fa5-a5ee-0024436915e3.png)

**Reply by alecthomas** [reply]
Aah. You're probably on a tagged release. Use `!` instead for now.

**Reply by alecthomas** [reply]
Yeah sorry about that, v2 is still in alpha so I'm still changing a few things. This construct was originally `!<expr>` but this turned out to be ambiguous so I had to change it to `~`. The README reflects this change but I haven't tagged a new alphs release yet. Details in the original commit (5c4f5190a26bb2e0f17c978e6ebe6004e9be7b36).

You could also just update to `@master` and `~` should work.

**Reply by leaanthony** [reply]
Thanks Alec. I'll try that 👍

**Reply by leaanthony** [reply]
Looks like it doesn't work. 
Input: `uuid(76eceacb-0462-4d94-ac83-423a6793775e)`
Rule: `parser:"'uuid' '(' @!')' ')' "`
Error: `exponent has no digits`

Looks like that's getting raised by the Go Scanner, trying to parse a number: https://github.com/golang/go/blob/master/src/go/scanner/scanner.go#L513

Also tried adding the String and Char type to see if it would stop it trying to parse the data as a number, to no avail. Any suggestions other than recreating the whole lexer? Cheers

**Reply by alecthomas** [reply]
I think you will almost certainly have to roll your own lexer, but it's not too onerous. [Here's](https://github.com/alecthomas/participle/blob/master/lexer/stateful_test.go#L429-L437) a simple example from the tests.
